File: jssrc/State.js
/**
 *  $Id: State.js 2420 2007-11-10 05:15:52Z nick $
 */
function State(string, temporary) {
	this.stateName = string;
	this.temporary = temporary;
	this.entryActions = Array();
	this.exitActions = Array();
	this.transitions = new Set();
	if(temporary == null) {
		this.temporary = false;
	}
}

State.prototype = {
	
	addTransition : function(pTransition) {
		if(pTransition.getFrom() !== this) {
			throw 'Programming error: Transitions supposed origin: '+ pTransition.getFrom() + ' is not the same as ' + this;
		}
		this.transitions.add(pTransition);
	},

	isTemporary : function() {
		return this.temporary;
	},

	addEntryAction : function(action) {
		this.entryActions.push(action);
	},
	
	addExitAction : function(action) {
		this.exitActions.push(action);
	},

	/**
	 * This outputs in GraphViz 'dot' notation suitable for OmniGraffle and GraphViz
	 */
	toString : function() {
		var diagraph = this.stateName.replace(/\s/g, '') + '[shape=';
		if(this.temporary) {
			diagraph += 'circle,';
		} else {
			diagraph += 'doublecircle,';
		}
		diagraph += 'color=black,'
		var label = this.stateName.replace(/\s/g, '\\n')
		diagraph += 'label="'+label+'"];\n';
		var tr = this.transitions.toArray();
		for(var i = 0; i < tr.length; i++) {
			diagraph += tr[i];
		}
		diagraph += '\n';
		return diagraph;
	},

	getTransitions : function(event) {
		var eventTransitions = Array();
		var ta = this.transitions.toArray();
		for(var i = 0; i < ta.length; i++) {
			if(ta[i].getEvents().contains(event)) {
				eventTransitions.push(ta[i]);
			}
		}
		return eventTransitions;
	},

	getAllTransitions : function() {
		return this.transitions.toArray();
	},

	getEntryActions : function() {
		return this.entryActions;	
	},

	getExitActions : function() {
		return this.exitActions;
	},

	getStateName : function() {
		return this.stateName;	
	}
};

File: src/com/cortexity/fsm/State.java
package com.cortexity.fsm;

import java.util.Collection;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public class State {

   private String name;

   private final Map<StateMachineEvent, Set<StateTransition>> transitions = new HashMap<StateMachineEvent, Set<StateTransition>>();

   private final Collection<StateMachineAction> entryActions = new HashSet<StateMachineAction>();

   private final Collection<StateMachineAction> exitActions = new HashSet<StateMachineAction>();

   private boolean temporary;

   public State(String string) {
      this(string, false);
   }

   public State(String string, boolean temporary) {
      name = string;
      this.temporary = temporary;
   }

   public String getName() {
      return name;
   }

   protected void addTranstion(StateTransition transition) {
      Set<StateMachineEvent> events = transition.getEvents();
      for (StateMachineEvent event : events) {
         Set<StateTransition> set = transitions.get(event);
         if (null == set) {
            set = new HashSet<StateTransition>();
         }
         set.add(transition);
         transitions.put(event, set);
      }
   }

   Collection<StateTransition> getTransitions(StateMachineEvent event) {
      Set<StateTransition> trans = transitions.get(event);
      if (null == trans) {
         trans = new HashSet<StateTransition>();
         transitions.put(event, trans);
      }
      return trans;
   }

   Collection<StateMachineAction> getEntryActions() {
      return entryActions;
   }

   @Override
   public String toString() {
      return name;
   }

   public void addEntryAction(StateMachineAction action) {
      this.entryActions.add(action);
   }

   public void addExitAction(StateMachineAction action) {
      this.exitActions.add(action);
   }

   Collection<StateMachineAction> getExitActions() {
      return this.exitActions;
   }

   public boolean isTemporary() {
      return temporary;
   }
}
